### ISSUE TITLE
ThreadSanitizer: data race /home/threadpool/includes/ThreadDispatcher.hpp:37 in ThreadPool::ThreadDispatcher<5ul>::ThreadDispatcher()

### ISSUE THREAD

**Issue by JohnSmithBH84** [farpeter98/threadpool#1]
I’ve build threadpool with Thread Satitizer (TSan) and it detected a data race in ThreadDispatcher.hpp after launching threadpool/build/threadpool. I'm not sure whether this is an actual issue, or whether threadpool' code is too sophisticated for TSan to understand. Do you guys have any idea?

# Project version
threadpool: (main 0d7b79fc53814c3dd3c8dd3b00e7153dad4d7780)
clang: 19.1.0 (x86_64-unknown-linux-gnu)

# Operating system
Ubuntu 22.04.4 LTS
Linux 5.15.0-125-generic #135-Ubuntu SMP Fri Sep 27 13:53:58 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

# I did this
Added the next rows into CMakeLists.txt to build with Thread Sanitizer.
```
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=thread -fno-omit-frame-pointer")
set(CMAKE_LINKER_FLAGS "${CMAKE_LINKER_FLAGS} -fsanitize=thread -fno-omit-frame-pointer")
set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```
Configured and build the project. 
```
mkdir build & cd build
cmake -DCMAKE_BUILD_TYPE=DEBUG -Wno-dev ..
cmake --build . -- -j16
```
Launched executable without arguments.
`./FlexFringe/build/runtests
`
# I expected the following
No data race reported by TSan

# I got the following
```
==================
WARNING: ThreadSanitizer: data race (pid=96)
  Write of size 1 at 0x7ffdd998b960 by main thread:
    #0 ThreadPool::ThreadDispatcher<5ul>::ThreadDispatcher() /home/threadpool/includes/ThreadDispatcher.hpp:37 (threadpool+0x29b5d)
    #1 main /home/threadpool/main.cpp:31 (threadpool+0x256fb)

  Previous read of size 1 at 0x7ffdd998b960 by thread T6:
    #0 ThreadPool::ThreadDispatcher<5ul>::DispatcherTask() /home/threadpool/includes/ThreadDispatcher.hpp:81 (threadpool+0x2bf58)
    #1 void std::__invoke_impl<void, void (ThreadPool::ThreadDispatcher<5ul>::*)(), ThreadPool::ThreadDispatcher<5ul>*>(std::__invoke_memfun_deref, void (ThreadPool::ThreadDispatcher<5ul>::*&&)(), ThreadPool::ThreadDispatcher<5ul>*&&) /usr/include/c++/12/bits/invoke.h:74 (threadpool+0x3827f)
    #2 std::__invoke_result<void (ThreadPool::ThreadDispatcher<5ul>::*)(), ThreadPool::ThreadDispatcher<5ul>*>::type std::__invoke<void (ThreadPool::ThreadDispatcher<5ul>::*)(), ThreadPool::ThreadDispatcher<5ul>*>(void (ThreadPool::ThreadDispatcher<5ul>::*&&)(), ThreadPool::ThreadDispatcher<5ul>*&&) /usr/include/c++/12/bits/invoke.h:96 (threadpool+0x3806c)
    #3 void std::thread::_Invoker<std::tuple<void (ThreadPool::ThreadDispatcher<5ul>::*)(), ThreadPool::ThreadDispatcher<5ul>*> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) /usr/include/c++/12/bits/std_thread.h:252 (threadpool+0x37f0c)
    #4 std::thread::_Invoker<std::tuple<void (ThreadPool::ThreadDispatcher<5ul>::*)(), ThreadPool::ThreadDispatcher<5ul>*> >::operator()() /usr/include/c++/12/bits/std_thread.h:259 (threadpool+0x37d64)
    #5 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (ThreadPool::ThreadDispatcher<5ul>::*)(), ThreadPool::ThreadDispatcher<5ul>*> > >::_M_run() /usr/include/c++/12/bits/std_thread.h:210 (threadpool+0x37a92)
    #6 <null> <null> (libstdc++.so.6+0xd44a2)

  Location is stack of main thread.

  Location is global '<null>' at 0x000000000000 ([stack]+0x1f960)

  Thread T6 (tid=103, finished) created by main thread at:
    #0 pthread_create ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:1001 (libtsan.so.2+0x5e686)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xd4578)
    #2 ThreadPool::ThreadDispatcher<5ul>::ThreadDispatcher() /home/threadpool/includes/ThreadDispatcher.hpp:36 (threadpool+0x29b4b)
    #3 main /home/threadpool/main.cpp:31 (threadpool+0x256fb)

SUMMARY: ThreadSanitizer: data race /home/threadpool/includes/ThreadDispatcher.hpp:37 in ThreadPool::ThreadDispatcher<5ul>::ThreadDispatcher()
==================

```

**Comment by farpeter98** [comment]
Hey, first of all, thanks for testing this small project!

Looking at the TSan log it seems like the issue is around the construction of `ThreadDispatcher` I think this is caused by the `dispatcherThread` being started before the `isRunning` flag is initialized due to their member declaration order, as well as the latter not being atomic, thus creating a possible data race where `DispatcherTask` might start while `isRunning` is in an undefined state.

`PooledThread` uses a similar mechanism but the declaration order is correct there so only the type needs to be changed, I did this quick modification for now, although as I was reading through the code again I noticed a few other issues around how `PooledThread` manages its stored task, namely that operations on `task` which is of type `std::optional<std::function<void ()>>` are done without any sort of synchronization mechanism even though operations on it are not atomic, so that could cause other data races.

That being said this was a small learning project I don't really plan on continuing, so I'm fine with leaving this second matter as it is for now, a 'proper' solution should probably be using some sort of queue instead of an optional anyhow, regardless, thanks again for the feedback!
